### PR TITLE
feat: add skipCrds flag for helm charts

### LINKED ISSUE #6252: Add ability to disable installation of v3 helm chart CRDs
# Summary

Currently with Argo CD when you define an Application declaratively to sync a V3 helm chart there is no way to disable the installation of CRDs.

For example, in the aws-load-balancer-controller helm chart the CRDs are defined in the [crds](https://github.com/aws/eks-charts/tree/master/stable/aws-load-balancer-controller/crds) directory and there is no way using a helm value to template this directory. The following application will force the user to sync the CRDs:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  finalizers:
    - resources-finalizer.argocd.argoproj.io
  name: aws-load-balancer-controller
  namespace: argocd
spec:
  destination:
    namespace: default
    server: https://kubernetes.default.svc
  project: someproject
  source:
    chart: aws-load-balancer-controller
    repoURL: https://aws.github.io/eks-charts
    targetRevision: 1.1.4
```

# Possible Workaround

Currently to get around this limitation I am able to set the `.spec.source.helm.version` property to `v2` to cause ArgoCD to not sync the contents of the `crds` directory, like this:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  finalizers:
    - resources-finalizer.argocd.argoproj.io
  name: aws-load-balancer-controller
  namespace: argocd
spec:
  destination:
    namespace: default
    server: https://kubernetes.default.svc
  project: someproject
  source:
    helm:
      version: v2 # This causes v2 templating to be used with ArgoCD
    chart: aws-load-balancer-controller
    repoURL: https://aws.github.io/eks-charts
    targetRevision: 1.1.4
```

## Update

Setting `version: v2` will only work on helm charts which have `apiVersion: v1` in Chart.yaml. So this isn't really a work around because CRDs would be managed as basic templates with `apiVersion: v1`.

# Motivation

In the kubernetes clusters which I manage I have a single ArgoCD Application which manages the installation of all CRDs since they are cluster-wide resources. We do not allow individual applications to install CRDs on their own because they could potentially conflict with other application CRDs. 

# Proposal

Add `.spec.source.helm.installCRDs` parameter to the Application CRD to allow individual applications to specify whether or not their CRDs should be synced.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  finalizers:
    - resources-finalizer.argocd.argoproj.io
  name: aws-load-balancer-controller
  namespace: argocd
spec:
  destination:
    namespace: default
    server: https://kubernetes.default.svc
  project: someproject
  source:
    helm:
      installCRDs: false # add this property
    chart: aws-load-balancer-controller
    repoURL: https://aws.github.io/eks-charts
    targetRevision: 1.1.4
```

### PR DESCRIPTION
**Pull request by patst**
Closes #6252

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 

This PR allows to skip the CRD creation for helm charts by providing a flag `skipCrds` .

I removed the `additionalTemplateArgs` array in the template options. There was only on entry and it was conflicting in turning crd application off.
Helm provides a `skip-crds` command line argument. But argo-cd uses `helm template` to generate rendered YAML files which are applied then. Imho `skip-crds` only works for `helm install`. 

I am unsure if a better name for the flag would be `include-crds` which is true by default. This would reflect the current behaviour.

From an end user perspective used to the `helm install` command the `skip-crds` flag is what we want to have.

### CODE REVIEW

**Comment by jplanza-gr** [comment]
The folks over at `prometheus-community` decided to take the ball home and close the playground; any help you can provide here would be greatly appreciated.

https://github.com/prometheus-community/helm-charts/pull/1510#issuecomment-1007554251

**Comment by sathieu** [comment]
@jplanza-gr Yes. But the correct fix is more #2267.

**Comment by jplanza-gr** [comment]
> @jplanza-gr Yes. But the correct fix is more #2267.

Thanks, I'll keep an eye on that issue.

**Comment by poksiala** [comment]
Even though the correct fix for kube-prometheus-stack case would be different I feel like this PR in itself would have value as well. I myself like to manage CRDs separately from applications and would love to see this merged. Some other valid uses have been raised in https://github.com/argoproj/argo-cd/issues/4723

**Comment by patst** [comment]
> The folks over at `prometheus-community` decided to take the ball home and close the playground; any help you can provide here would be greatly appreciated.
> 
> [prometheus-community/helm-charts#1510 (comment)](https://github.com/prometheus-community/helm-charts/pull/1510#issuecomment-1007554251)

I agree with @poksiala . This is not completely related to the prometheus helm chart.
I like the writeup at https://github.com/helm/community/blob/f9e06c16d89ccea1bea77c01a6a96ae3b309f823/architecture/crds.md which covers all pros and cons of different ways of handling CRDs. 

I would like to have a separate application to maintain the CRD version and therefore have the possibility to skip the installation via the helm3 `crd` folder.
This does not work with ArgoCD at the moment because under the hood the helm charts are *installed* using `helm template --include-crds` + `kubectl apply`. 

The `include-crds` is not configurable at the moment which changes with this PR.

File: util/helm/cmd.go [hunk 1]
```diff
@@ -330,8 +331,8 @@ func (c *Cmd) template(chartPath string, opts *TemplateOpts) (string, error) {
 	for _, v := range opts.APIVersions {
 		args = append(args, "--api-versions", v)
 	}
-	if c.HelmVer.additionalTemplateArgs != nil {
-		args = append(args, c.HelmVer.additionalTemplateArgs...)
+	if !opts.SkipCrds {
```
Comment (alexmt) [inline util/helm/cmd.go:334]: Nice, thank you for keeping backward compatible behavior!

**Review by alexmt — COMMENTED** [review]
LGTM. Thank you!

**Review by alexmt — CHANGES_REQUESTED** [review]
E2E tests have caught an issue: we still support helm2 . The `--skip-crds` flag should be added only for Helm3. @patst , can you please fix it.

**Comment by patst** [comment]
> E2E tests have caught an issue: we still support helm2 . The `--skip-crds` flag should be added only for Helm3. @patst , can you please fix it.

@alexmt : thanks for your feedback.

I missed the one condition which makes sure `-include_crds` is not passed for helm2 if `skip-crds`  is unset. I think this is fixed now. Would be great to have the E2E tests run again

**Comment by patst** [comment]
the Lint docs failed due to an unrelated issue because `https://grafana.apps.argoproj.io/` wasn't reachable at the moment:

```
> Checking links in docs/operator-manual/metrics.md
> Will allow duplicate links
> Will allow redirects
> Connection timeout = 3s
> White list links matching: 10.97.164.88, 192.168.0.20, argocd.example.com, api.github.com/user, cd.apps.argoproj.io, docker-build, docker-build:443, git.example.com, git.example.com:443, github.com/argoproj/another-private-repo, github.com/argoproj/my-private-repository, github.com/argoproj/other-private-repo, github.com/argoproj/private-repo, github.com/otherproj/another-private-repo, ksonnet.io, raw.githubusercontent.com/argoproj/argo-cd, repo.example.com, repo.example.com:443, server.example.com, kubernetes.default.svc, kubernetes.default.svc:443, localhost:4000, localhost:6443, localhost:8080, localhost:8085, mycluster.com, storage.googleapis.com, ui.argocd.yourorganization.net, ui.argocd.yourorganization.net:443, your-kubernetes-cluster-addr, yourorganization.oktapreview.com, yourorganization.oktapreview.com:443, example-OIDC-provider.com, argocd-dex-server:5556, ghe.example.com, proxy-server-url:8888, keycloak.example.com, argocd.myproject.com, argocd.apps.domain.com, k8sou.apps.192-168-2-144.nip.io, your.argoingress.address, your.domain, external.path.to.argocd.io, my-argo-cd-url, my-login-url, login.microsoftonline.com/xxxxx, accounts.google.com/o/saml2/idp?idpid=Abcde0, accounts.google.com/o/saml2?idpid=Abcde0, sso-url, google-entity-id, github.com/argoproj/argo-cd/manifests/crds, example.com, form.example.com, grafana.example.com, 10.5.39.39, chat.googleapis.com/v1/spaces/, mattermost.example.com, my-grafana.com, github.my-company.com, 1.2.3.4, 2.4.6.8, 9.8.7.6, ghe.example.com, 12.34.567.89, 192.168.99.100:8443, github.com/yourghuser/argo-cd, github.com/argoproj/argo-cd/releases/download/, https://github.com/hayorov/helm-gcs.git; 
> Will not save results
Links to check: 2, 0 white listed
  1. https://github.com/argoproj/argo-cd/blob/master/examples/dashboard.json 
  2. https://grafana.apps.argoproj.io 
Checking URLs: ?✓
Issues :-(
> Links 
  1. [L137]  https://grafana.apps.argoproj.io Failed to open TCP connection to grafana.apps.argoproj.io:443 (Connection refused - connect(2) for "grafana.apps.argoproj.io" port 443) 
```

**Review by alexmt — APPROVED** [review]

**Pull request merged**